### ISSUE TITLE
feature request: expose interfaces that other classes can implement

### ISSUE THREAD

**Issue by pschiffmann** [google/built_collection.dart#107]
I have written a class [`Superset`](https://github.com/pschiffmann/indexed-set.dart/blob/master/lib/src/superset.dart) that implements all methods of `BuiltSet`. I would like to actually write `implements BuiltSet` (for the static analysis hints during development), but that would break [this](https://github.com/google/built_collection.dart/blob/c6fdd6d30338f982ebad88506ef213fa67be80e7/lib/src/set/set_builder.dart#L49) line in your code.

Related(?) issue: https://github.com/dart-lang/site-www/issues/382

Btw: I can contribute a pull request for this (and the other issues I opened recently). I guess you assign yourself to an issue if you started working on it?

**Comment by davidmorgan** [comment]
Thanks! Is the suggested fix to check .runtimeType instead of "is"?

PRs are welcome; I generally get to built_value/built_collection issues quite quickly if they're blocking someone, but non-critical issues wait until I have some spare time, which can take a while :)

**Comment by pschiffmann** [comment]
Since a runtimeType check will break for subclasses (`class MyBuiltSet extends BuiltSet`), I was going to suggest the same approach the [quiver](https://pub.dartlang.org/packages/quiver) authors use for their classes: Provide an abstract class that defines the interface, then provide a default implementation of that interface in a separate class, then add a factory constructor to the interface class that returns an instance of the default implementation.
But now that you mention it, the runtimeType check might be sufficient in this case, because built types must never be subclassed. (if I recall correctly)

Which approach do you prefer?

**Comment by davidmorgan** [comment]
Yeah, I don't particularly want to encourage subclasses; if there's something people want to do along those lines I hope they would get in touch :)

So I think checking runtimeType sounds reasonable.

**Comment by pschiffmann** [comment]
There's a problem with runtimeType: `new BuiltList<int>().runtimeType == BuiltList` returns false, because the runtimeType is `BuiltList<int>`. Both the [factory constructor](https://github.com/google/built_collection.dart/blob/966de498b2de397db0e7c5f12073cb2dc64ca5fc/lib/src/list/built_list.dart#L30) and [`ListBuilder.replace`](https://github.com/google/built_collection.dart/blob/966de498b2de397db0e7c5f12073cb2dc64ca5fc/lib/src/list/list_builder.dart#L49) don't have access to the exact expected type.

I could work around this with
`if (iterable.runtimeType == new BuiltList<E>().runtimeType)`
but that involves an unnecessary allocation.

Alternatively, I could do (example for BuiltList factory)
```
try {
  return iterable as BuiltList<E>;
on CastError {
  return new BuiltList<E>._copyAndCheck(iterable);
}
```
but catching an `Error` goes against the Dart style guide.

Your thoughts on this?

**Comment by davidmorgan** [comment]
Tricky. We could parse `runtimeType.toString`, but I'm not sure that would be any faster. I'll see if anyone on the languages team has suggestions.

How does it work out if you try to implement those private members? Not hugely pretty, but might be a short term solution.

**Comment by pschiffmann** [comment]
> How does it work out if you try to implement those private members?
Impossible because protected members are only visible to the same library. If I add a `_set` property to my class, `SetBuilder` still can't access it and will throw `NoSuchMethodError`.
> Not hugely pretty, but might be a short term solution.
It's not urgent, I'm probably the only one who will use this package in the near future anyways, and I don't need to create a regular BuiltSet with the contents of my own IndexedSet. I'd just prefer to publish the package when it doesn't have these pitfalls :)

**Comment by davidmorgan** [comment]
Ah. Forgot that bit. Heh.

The language team says to go ahead and create a proper interface, as in your original request :) ... I was worried about possible performance implications but apparently there aren't any.

Seems reasonable.

**Comment by davidmorgan** [comment]
They also suggested this pattern to avoid having to write the whole public API twice:

```
abstract class BuiltSometing<T> ... {
   factory BuiltSomething(...) = _BuiltSomething<T>;
   BuiltSomething._(...);
   ... actual implementation ...
}

class _BuiltSomething<T> extends BuiltSomething<T> {
  _BuiltSomething(...) : super._(...);
}
```

The implementation stays in the public class, but is only ever used concretely in the private subclass.

**Comment by pschiffmann** [comment]
OK. I can do that, if you want.

Do you have suggestions/preferences in which files the interfaces should go, and how to rename the existing classes? If not, I'd start with one class, then ask you for a review before I implement the others.

**Comment by davidmorgan** [comment]
Sounds good, please do :)

I'd keep BuiltList in the same file, and add _BuiltList as the implementation class in that same file. How does that sound?

Thanks.

**Comment by pschiffmann** [comment]
I just remembered that #76 is still open, too. If the default BuiltList becomes a wrapper of arbitrary Lists, should it maybe be called `WrappingBuiltList` and be part of the public package API? Like this:

```
abstract class BuiltSometing<T> ... {
   factory BuiltSomething(...) = WrappingBuiltSomething<T>;
   BuiltSomething._(...);
   ... actual implementation ...
}

class WrappingBuiltSomething<T> extends BuiltSomething<T> {
  WrappingBuiltSomething(...) : super._(...);
  WrappingBuiltSomething.using(List<T> base) { ... }
}

**Comment by davidmorgan** [comment]
If it's public then we have the same problem that someone might implement it :) ... we can use different factory constructors on BuiltSomething to expose different implementations that might show up later, rather than needing to expose different types.

**Comment by davidmorgan** [comment]
Published as 2.0.0. Now to fix built_value :) ... should be able to get a fix out tomorrow.

Thanks!
